extensions/amp-video/0.1/flexible-bitrate.js

```diff
@@ -0,0 +1,257 @@
+   * the sources closest matching the acceptable bitrate are in front.
+   * @param {!Element} video
+   */
+  sortSources_(video) {
+    const sources = toArray(childElementsByTag(video, 'source'));
+    // Ensure each element has the bitrate_ property
+    sources.forEach((source) => {
+      if (source.bitrate_) {
+        return;
+      }
+      const bitrate = source.getAttribute('data-bitrate');
+      source.bitrate_ = bitrate ? parseInt(bitrate, 10) : 0;
+    });
+    sources.sort((a, b) => {
+      // Biggest first, bitrates above threshold to the back
+      return (
+        this.getBitrateForComparisson_(b) - this.getBitrateForComparisson_(a)
+      );
+    });
+    sources.forEach((source) => {
+      video.appendChild(source);
+    });
+  }
+
+  /**
+   * Returns the bitrate of a source in such a way that bitrates that are higher
+   * than the current acceptable bitrate rank lowest.
+   * @param {!Element} source
+   * @return {number}
+   * @private
+   */
+  getBitrateForComparisson_(source) {
+    let rate = source.bitrate_;
+    if (rate > this.getAcceptableBitrate_()) {
+      // We make bitrates higher than the acceptable rate negative. That means,
+      // the higher they are, the further they go in the back.
+      rate *= -1;
+    }
+    return rate;
+  }
+
+  /**
+   * Returns true if any sources of the video have a lower bitrate than the given
+   * bitrate.
+   * @param {!Element} video
+   * @param {number} bitrate
+   * @return {boolean}
+   * @private
+   */
+  hasLowerBitrate_(video, bitrate) {
+    const lowerBitrateSource = sources(video, (source) => {
+      return source.bitrate_ < bitrate;
+    });
+    return !!lowerBitrateSource;
+  }
+
+  /**
+   * Switches the video to use a lower bitrate if available.
+   * This should be called if the video is currently in waiting mode.
+   * @param {!Element} video
+   * @param {number} currentBitrate
+   * @private
+   */
+  switchToLowerBitrate_(video, currentBitrate) {
+    if (!this.hasLowerBitrate_(video, currentBitrate)) {
+      dev().fine(TAG, 'No lower bitrate available');
+      return;
+    }
+    const {currentTime} = video;
+    video.pause();
+    this.sortSources_(video);
+    video.load();
+    listenOnce(video, 'loadedmetadata', () => {
+      // Restore currentTime after loading new source.
+      video.currentTime = currentTime;
+      video.play();
+      dev().fine(TAG, 'Playing at lower bitrate %s', video.currentSrc);
+    });
+  }
+}
+
+/**
+ * Calls the callback if the video goes into waiting state and does not
+ * emerge from it within a short amount of time.
+ * @param {!Element} video
+ * @param {function()} callback
+ */
+function onNontrivialWait(video, callback) {
+  listen(video, 'waiting', () => {
+    let timer = null;
+    const unlisten = listenOnce(video, 'playing', () => {
+      clearTimeout(timer);
+    });
+    timer = setTimeout(() => {
+      unlisten();
+      callback();
+    }, 50);
```

Comment:
I wish I had better metrics available to pick a threshold here. We use 100ms to show the loading spinner on the page, and 250ms to count it as a "rebuffer". Looking at our dashboards and rebuffer data, 250 is definitely too high and we should update it, but 50ms sounds low.